### DISCUSSION TITLE
Android egui app

### DISCUSSION THREAD

**Discussion by devourers** [emilk/egui#2053] [Q&A]
I would like to apologize in advance, because this is probably a very noob question, but I am fairly new to Rust and have little to no idea how android apps work.
My goal is to build a Rust app for android, but I would like to refrain from using android studio and just code in pure rust. I recently found out about `cargo-apk`, and using it wanted to build my app for phone. As a start I was trying to just run hello-world example from egui's repo, by adding `#[cfg_attr(target_os = "android", ndk_glue::main(backtrace = "on"))]` to `main` function, and running `cargo apk run --lib` with started emulator.
Here is the whole code:
```
use eframe::egui;

struct MyApp {
    name: String,
    age: u32,
}

impl Default for MyApp {
    fn default() -> Self {
        Self {
            name: "Arthur".to_owned(),
            age: 42,
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.heading("My egui Application");
            ui.horizontal(|ui| {
                ui.label("Your name: ");
                ui.text_edit_singleline(&mut self.name);
            });
            ui.add(egui::Slider::new(&mut self.age, 0..=120).text("age"));
            if ui.button("Click each year").clicked() {
                self.age += 1;
            }
            ui.label(format!("Hello '{}', age {}", self.name, self.age));
        });
    }
}

#[cfg_attr(target_os = "android", ndk_glue::main(backtrace = "on"))]
fn main() {
    let options = eframe::NativeOptions::default();
    eframe::run_native(
        "My egui App",
        options,
        Box::new(|_cc| Box::new(MyApp::default())),
    );
}
```
The programm starts, it also even outputs `prinln!`s to log, but then the log gives me `thread '<unnamed>' panicked at 'Cannot get the native window, it's null and will always be null before Event::Resumed and after Event::Suspended. Make sure you only call this function between those events.'`, which I consider that my rust code died and that's it.
Can someone explain to me if it's 

1. me doing something wrong.
2. Rust android develpoment with egui requiring something else
3. Rust egui not working on android as is.

Thanks in advance!

**Comment by Parth** [comment]
https://github.com/emilk/egui/pull/1952 was merged about 1 month before your example ran, I believe these changes are still not released. Try to update your toml file to point to egui master and see if android works as expected. Let me know if it does, I'm personally curious.

@rib am I matching up the error above with your work correctly?

**Reply by rib** [reply]
Yeah running with eframe on Android requires those changes. I guess they should be release by now though - might be wrong.

That is the kind of error I would expect though if not running with the changes above.

For egui-eframe in particular maybe you can take a look at this android-activity example: https://github.com/rib/android-activity/tree/main/examples/agdk-eframe

That example is based on `GameActivity` instead of `NativeActivity` so it needs Gradle to build.

Personally I wouldn't try to hard to only use 100% Rust if you want to target Android. Since the platform is fundamentally JVM based you can quite quickly hit limits in what's possible if you only use native code.

For testing then the above `egui-eframe`  example should probably also work with `NativeActivity` ok too and then you'd be able to build with `cargo apk`. The main change would just be to modify the `Cargo.toml` to use the "native-activity" feature instead of "game-activity".

**Reply by m00nwtchr** [reply]
I have a similar issue to OP.
I tried using the `master` version of egui and now I'm getting:
```
RustStdoutStderr: thread '<unnamed>' panicked at 'Supplied timeout is too large: TryFromIntError(())', /home/m00n/.local/share/cargo/registry/src/github.com-1ecc6299db9ec823/ndk-0.7.0/src/looper.rs:167:18
```

**Reply by rib** [reply]
That vaguely sounds like an issue I've seen with egui before due based on how your code handles timeouts based on egui's integration output data. I vaguely recall that you need to check that if egui is asking for a ridiculously long (infinite) timeout then you need to make sure you don't ask Winit to wait for that timeout, rather you should just ask winit to wait for new events without specifying any timeout.

**Reply by rib** [reply]
E.g. I used to have some code like this for mapping `full_output.repaint_after` to Winit control_flow state before I switched to using eframe that hides this:

```rust
    // This seems like some pretty funky API design :/
    //
    // Winit should probably have some kind of timers API to remove the need for
    // WaitUntil and there should maybe be an explicit .quit() API for the event
    // loop that would avoid this awkward control_flow precedence issue.
    //
    // It also seems pretty odd for Egui to specify a wait _duration_ when it
    // doesn't know when we will start waiting - surely they should specify
    // an optional deadline instant instead.
    if *control_flow != winit::event_loop::ControlFlow::Exit {
        *control_flow =
            if full_output.repaint_after.is_zero() || !emulator_ui.paused {
                window.request_redraw();
                winit::event_loop::ControlFlow::Poll
            } else if let Some(repaint_after_instant) =
                std::time::Instant::now().checked_add(full_output.repaint_after)
            {
                // if repaint_after is something huge and can't be added to Instant,
                // we will use `ControlFlow::Wait` instead.
                // technically, this might lead to some weird corner cases where the user *WANTS*
                // winit to use `WaitUntil(MAX_INSTANT)` explicitly. they can roll their own
                // egui backend impl i guess.
                winit::event_loop::ControlFlow::WaitUntil(repaint_after_instant)
            } else {
                winit::event_loop::ControlFlow::Wait
            };
    }
```

I'm fairly sure that when I didn't handle the case where `checked_add` could fail then I hit an error similar to the one above

**Reply by m00nwtchr** [reply]
I am using eframe though🤔

**Reply by rib** [reply]
maybe it's worth double checking in case something changed about how eframe handles this mapping - any maybe it's now not handling this case properly?

**Reply by rib** [reply]
from skimming that code in eframe it looks like it uses checked_add, so then maybe it's something else entirely unrelated - that's just what popped into my mind from seeing a similar error with egui in the past.

**Reply by rib** [reply]
the fact that the panic is in `src/looper.rs:167` does generally make me think it's still likely to be an issue with how the `repaint_after` time is being mapped to Winit a `control flow` since that would potentially result in passing a timeout through to the Looper to poll with a timeout and it's liable to panic if it can't be converted into an `int` for the Looper API

**Reply by rib** [reply]
I think in some other app I have used an arbitrary threshold for this instead of relying on `checked_add` failing. `checked_add` might still accept ridiculously large durations that still won't be possible to convert into a valid timeout when polling later

**Reply by rib** [reply]
Yeah e.g. in this app I just have an arbitrary 1 second threshold:

https://github.com/rib/bluey/blob/5f351c50e6f396bc4fef9c050599f9668c63b339/bluey-ui/src/lib.rs#L144

**Reply by rib** [reply]
Maybe eframe should use a threshold along those lines instead of relying on `checked_add`

**Comment by Parth** [comment]
I'm approaching this from the iOS side of things but will ultimately need to answer these questions on android. Other things I would look into are:
+ https://github.com/rib/android-activity/tree/main/examples/agdk-egui
+ https://github.com/frgomes/egui-miniquad-demo

**Comment by expenses** [comment]
I've been working on writing an eframe app for android. It's pretty simple, things mostly just work as-is with eframe: https://github.com/expenses/android-eframe/commit/a65a3f0ac5eba2dbc08fdb4220f76d8e9ee615c6.

**Reply by expenses** [reply]
Biggest problem so far is that the soft keyboard doesn't pop up when you focus on input text box. You also need to setup a `TopBottomPanel` or to pad out the screen at the top (where the camera notch and icons are).

**Reply by expenses** [reply]
https://github.com/rust-windowing/winit/pull/2993 is a good start on the soft keyboard thing, but I'm blocked from trying it out because of https://github.com/emilk/egui/issues/4709.

**Reply by DougAnderson444** [reply]
@expenses, I saw from https://github.com/rust-mobile/ndk/tree/master?tab=readme-ov-file that cargo-apk	https://github.com/rust-mobile/cargo-apk is ⛔ deprecated in favour of [xbuild](https://github.com/rust-mobile/xbuild). Does it matter which one we use?

**Reply by DougAnderson444** [reply]
For those following the soft keyboard issue, it's now tracking at https://github.com/rust-windowing/winit/pull/3787 (superseding https://github.com/rust-windowing/winit/pull/2993)

**Reply by DougAnderson444** [reply]
As a temporary workaround for the soft keyboard issue, there is also: 
https://github.com/snail-with-tea/egui_virtual_keyboard

**Reply by lahwran** [reply]
is there an example I can try? would love to play with this
